### PR TITLE
DefaultTbLocalSubscriptionService WS single lock refactored by tenantId; TbEntityLocalSubsInfo performance optimizations for remove subscription

### PR DESCRIPTION
**Pull request by YevhenBondarenko**
## Pull Request description

Based on PR: https://github.com/thingsboard/thingsboard/pull/11388

## General checklist

- [ ] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [ ] [Labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#about-labels) that classify your pull request have been added.
- [ ] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [ ] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [ ] Description contains human-readable scope of changes.
- [ ] Description contains brief notes about what needs to be added to the documentation.
- [ ] No merge conflicts, commented blocks of code, code formatting issues.
- [ ] Changes are backward compatible or upgrade script is provided.
- [ ] Similar PR is opened for PE version to simplify merge. Crosslinks between PRs added. Required for internal contributors only.
  
## Front-End feature checklist

- [ ] Screenshots with affected component(s) are added. The best option is to provide 2 screens: before and after changes;
- [ ] If you change the widget or other API, ensure it is backward-compatible or upgrade script is present.
- [ ] Ensure new API is documented [here](https://github.com/thingsboard/thingsboard-ui-help)

## Back-End feature checklist

- [ ] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [ ] If new dependency was added: the dependency tree is checked for conflicts.
- [ ] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [ ] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.
- [ ] If new yml property was added: make sure a description is added (above or near the property).

### CODE REVIEW

File: application/src/main/java/org/thingsboard/server/service/subscription/DefaultTbLocalSubscriptionService.java [hunk 1]
```diff
@@ -207,18 +213,20 @@ public void addSubscription(TbSubscription<?> subscription, WebSocketSessionRef
...
     @Override
-    public void onSubEventCallback(EntityId entityId, int seqNumber, TbEntityUpdatesInfo entityUpdatesInfo, TbCallback callback) {
-        onSubEventCallback(entityId.getId(), seqNumber, entityUpdatesInfo, callback);
+    public void onSubEventCallback(TenantId tenantId, EntityId entityId, int seqNumber, TbEntityUpdatesInfo entityUpdatesInfo, TbCallback callback) {
+        onSubEventCallback(tenantId, entityId.getId(), seqNumber, entityUpdatesInfo, callback);
     }
 
-    public void onSubEventCallback(UUID entityId, int seqNumber, TbEntityUpdatesInfo entityUpdatesInfo, TbCallback callback) {
-        log.debug("[{}][{}] Processing sub event callback: {}.", entityId, seqNumber, entityUpdatesInfo);
+    public void onSubEventCallback(TenantId tenantId, UUID entityId, int seqNumber, TbEntityUpdatesInfo entityUpdatesInfo, TbCallback callback) {
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/subscription/DefaultTbLocalSubscriptionService.java:225]: this method may be private?

File: application/src/main/java/org/thingsboard/server/service/subscription/TbAttributeSubscription.java [hunk 2]
```diff
@@ -43,12 +43,8 @@ public TbAttributeSubscription(String serviceId, String sessionId, int subscript
     }
 
     @Override
-    public boolean equals(Object o) {
-        return super.equals(o);
+    protected boolean canEqual(final Object other) {
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/subscription/TbAttributeSubscription.java:46]: Why do you change this?

File: application/src/main/java/org/thingsboard/server/service/subscription/TbEntityLocalSubsInfo.java [hunk 3]
```diff
@@ -137,8 +135,30 @@ public TbEntitySubEvent remove(TbSubscription<?> sub) {
             return toEvent(ComponentLifecycleEvent.DELETED);
         }
         TbSubscriptionsInfo oldState = state.copy();
-        TbSubscriptionsInfo newState = new TbSubscriptionsInfo();
+
+        //copy unchanged state only
+        TbSubscriptionsInfo newState = state.copy();
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/subscription/TbEntityLocalSubsInfo.java:140]: Remove all code that is not related to the scope of changes. Create new PR and cover this with TESTS!!!

File: application/src/main/java/org/thingsboard/server/service/subscription/TbAttributeSubscription.java (hunk 2 above)
Comment (ashvayka) [reply application/src/main/java/org/thingsboard/server/service/subscription/TbAttributeSubscription.java:46]: Remove all code that is not related to the scope of changes. Create new PR and cover this with TESTS!!!

File: application/src/main/java/org/thingsboard/server/service/subscription/TbSubscriptionUtils.java [hunk 4]
```diff
@@ -82,10 +82,12 @@ public static ToCoreMsg toSubEventProto(String serviceId, TbEntitySubEvent event
         return ToCoreMsg.newBuilder().setToSubscriptionMgrMsg(msgBuilder).build();
     }
 
-    public static ToCoreNotificationMsg toProto(UUID id, int seqNumber, TbEntityUpdatesInfo update) {
+    public static ToCoreNotificationMsg toProto(TenantId tenantId, UUID id, int seqNumber, TbEntityUpdatesInfo update) {
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/subscription/TbSubscriptionUtils.java:85]: Remove all code that is not related to the scope of changes. Create new PR and cover this with TESTS!!!

File: application/src/main/java/org/thingsboard/server/service/subscription/TbSubscriptionsInfo.java [hunk 5]
```diff
@@ -20,25 +20,24 @@
...
 import java.util.Set;
-import java.util.concurrent.ConcurrentHashMap;
 
 /**
  * Information about the local websocket subscriptions.
  */
 @RequiredArgsConstructor
 @AllArgsConstructor
-@EqualsAndHashCode(exclude = {"seqNumber"})
+@EqualsAndHashCode
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/subscription/TbSubscriptionsInfo.java:30]: Remove all code that is not related to the scope of changes. Create new PR and cover this with TESTS!!!

File: application/src/main/java/org/thingsboard/server/service/subscription/TbTimeSeriesSubscription.java [hunk 6]
```diff
@@ -53,12 +53,8 @@ public TbTimeSeriesSubscription(String serviceId, String sessionId, int subscrip
     }
 
     @Override
-    public boolean equals(Object o) {
-        return super.equals(o);
+    protected boolean canEqual(final Object other) {
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/subscription/TbTimeSeriesSubscription.java:56]: Remove all code that is not related to the scope of changes. Create new PR and cover this with TESTS!!!

File: application/src/main/java/org/thingsboard/server/service/ws/notification/DefaultNotificationCommandsHandler.java [hunk 7]
```diff
@@ -242,7 +242,7 @@ public void handleMarkAllAsReadCmd(WebSocketSessionRef sessionRef, MarkAllNotifi
 
     @Override
     public void handleUnsubCmd(WebSocketSessionRef sessionRef, UnsubscribeCmd cmd) {
-        localSubscriptionService.cancelSubscription(sessionRef.getSessionId(), cmd.getCmdId());
+        localSubscriptionService.cancelSubscription(sessionRef.getSecurityCtx().getTenantId(), sessionRef.getSessionId(), cmd.getCmdId());
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/ws/notification/DefaultNotificationCommandsHandler.java:245]: Let's introduce a method `sessionRef.getTenantId()` and replace usages of `sessionRef.getSecurityCtx().getTenantId()` with it.
Also, please check in that method that securityCtx is not null (it will be null until user sends auth cmd), otherwise return sys tenant id.

**Review by ViacheslavKlimov — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/thingsboard/server/service/subscription/DefaultTbLocalSubscriptionService.java [hunk 8]
```diff
@@ -206,19 +212,26 @@ public void addSubscription(TbSubscription<?> subscription, WebSocketSessionRef
 
     @Override
     public void onSubEventCallback(TransportProtos.TbEntitySubEventCallbackProto subEventCallback, TbCallback callback) {
+        TenantId tenantId;
+        if (subEventCallback.getTenantIdMSB() == 0 && subEventCallback.getTenantIdLSB() == 0) {
+            tenantId = TenantId.SYS_TENANT_ID; //TODO: remove after release
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/subscription/DefaultTbLocalSubscriptionService.java:217]: :+1:
